### DISCUSSION TITLE
How to integrate elixir-tools

### DISCUSSION THREAD

**Discussion by Pilgrim1379** [VonHeikemen/lsp-zero.nvim#259] [Q&A] [answered]
There's a relatively new `elixir-tools` which to my understanding automatically sets up `credo language server` and `elixirls` for neovim. It's got a sample config on it's [project page](https://github.com/elixir-tools/elixir-tools.nvim#advanced-setup) and is currently not listed on the `lspconfig` website as supported. 

I'll like `lsp-zero` to handle it's setup. What's gotten me stuck is, going by their config, I cant figure out how to make it use `lsp-zero's` on_attach. I had quite a bit of functionality in my old on_attach which `lsp-zero` now provides. I just don't want to create a seperate `local an_attach` function, duplicating most of `lsp-zero` functionaly just for this one server.

**Comment by VonHeikemen** [comment]
Are you using the `v2.x` branch of lsp-zero? If you are the `on_attach()` in lsp-zero will just work. There is no need for lsp-zero to handle the setup of the LSP server.

**Reply by Pilgrim1379** [reply]
Yes I'm on the `v2.x` branch. If I understand you correctly, my current code is this:
```lua
-- Elixir Tools
local elixir = require("elixir")
local elixirls = require("elixir.elixirls")

elixir.setup {
  credo = {
    enable = true
  },
  elixirls = {
    enabled = true,
    cmd = "elixir-ls",
    settings = elixirls.settings {
      dialyzerEnabled = true,
      enableTestLenses = false,
    },
    on_attach = on_attach,
    capabilities = capabilities
  }
}
```

and your saying below should just work?
```lua
-- Elixir Tools
local elixir = require("elixir")
local elixirls = require("elixir.elixirls")

elixir.setup {
  credo = {
    enable = true
  },
  elixirls = {
    enabled = true,
    cmd = "elixir-ls",
    settings = elixirls.settings {
      dialyzerEnabled = true,
      enableTestLenses = false,
    },
  }
}
```
And should this come before `lsp.setup()` or after?

**Comment by VonHeikemen** [comment, marked as answer]
If you installed elixir-ls using mason.nvim, then it should be after.

You should still include the capabilities key in `elixirls`.

**Reply by Pilgrim1379** [reply]
Thanks. Actually I realised elixir-tools is trying to something similar to what rust-tools does so I addapted the [example setup](https://github.com/VonHeikemen/lsp-zero.nvim/blob/v2.x/doc/md/guides/quick-recipes.md#setup-with-rust-tools) and it's all good now.
